### PR TITLE
stft: Require return_complex to be passed explicitly for real input

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #86628
* __->__ #86724

This behavior has been deprecated since PyTorch 1.8 but this step of
the deprecation cycle was put on hold in #50102 waiting for JIT
upgraders functionality which doesn't seem to have panned out. I'd say
there has been more than enough of a deprecation period, so we should
just continue.

BC-breaking message:

`torch.stft` takes an optional `return_complex` parameter that
indicates whether the output should be a floating point tensor or a
complex tensor. `return_complex` previously defaulted to `False` for
real input tensors. This PR removes the default and makes
`return_complex` a required argument for real inputs. However, complex
inputs will continue to default to `return_complex=True`.

1.13
```python
>>> a = torch.rand(1024)
>>> _ = torch.stft(a, n_fft=128)
```

2.0
```python
>>> t = torch.rand(1024)
>>> _ = torch.stft(t, n_fft=128, return_complex=False)
```

### CODE REVIEW

File: test/test_spectral_ops.py [hunk 1]
```diff
@@ -1181,9 +1181,8 @@ def test_complex_stft_onesided(self, device):
     @skipCPUIfNoFFT
     def test_stft_requires_complex(self, device):
         x = torch.rand(100)
-        y = x.stft(10, pad_mode='constant')
-        # with self.assertRaisesRegex(RuntimeError, 'stft requires the return_complex parameter'):
-        #     y = x.stft(10, pad_mode='constant')
+        with self.assertRaisesRegex(RuntimeError, 'stft requires the return_complex parameter'):
```
Comment (mruberry) [inline test/test_spectral_ops.py:1184]: Nit: this could be an error input

File: torch/functional.py [hunk 2]
```diff
@@ -612,6 +612,15 @@ def stft(input: Tensor, n_fft: int, hop_length: Optional[int] = None,
             a real tensor with an extra last dimension for the real and
             imaginary components.
 
+            .. versionchanged:: 1.14.0
+               ``return_complex`` is now a required argument for real inputs,
+               as the default is being trainsitioned to ``True``.
```
Comment (mruberry) [inline torch/functional.py:617]: "trainsitioned" -> "transitioned"

**Review by mruberry — APPROVED** [review]
This seems reasonable to me -- @ngimel @albanD is there someone at Meta who should sanity-check the BC implications?

@peterbell10 take a look at my questions about the docs, however -- I think they can be streamlined

**Review by albanD — APPROVED** [review]
I think we should just land this!
Checking is so painful that it is easier to try and land it :) 

If it comes back to us we can go down the rabbit hole.

**Review by albanD — APPROVED** [review]

**Pull request closed without merging**